cborg/src/Codec/CBOR/Encoding.hs

```diff
@@ -95,10 +100,12 @@ data Tokens =
     | TkInt64    {-# UNPACK #-} !Int64        Tokens
 
     -- Bytes and string (type 2,3)
-    | TkBytes    {-# UNPACK #-} !B.ByteString Tokens
-    | TkBytesBegin                            Tokens
-    | TkString   {-# UNPACK #-} !T.Text       Tokens
-    | TkStringBegin                           Tokens
+    | TkBytes         {-# UNPACK #-} !B.ByteString Tokens
+    | TkBytesBegin                                 Tokens
+    | TkByteArray     {-# UNPACK #-} !BA.ByteArray Tokens
+    | TkString        {-# UNPACK #-} !T.Text       Tokens
+    | TkUtf8ByteArray {-# UNPACK #-} !BA.ByteArray Tokens
```

Comment:
I think the `TkByteArray` and `TkUtf8ByteArray` instructions should take an offset and length as well. This is the most general form and needed to be able to do non-copying for other higher level data structures (not least bytestring and vector).